### DISCUSSION TITLE
`read` equivalent of command substitution

### DISCUSSION THREAD

**Discussion by IlanCosman** [fish-shell/fish-shell#8564] [Q&A] [answered]
What's the equivalent of 
```fish
set foo (printf %s\n 1 2 3)
echo $foo[3] # 3
```

using `read`?

```fish
printf %s\n 1 2 3 | read --list foo
echo $foo[3] # Doesn't exist
```

I recall asking this on gitter a long time ago, but am now unable to find the answer.
I think Fabian said it was `read --null --list`, and that does seem to work, but I am wondering:
1. if there are any caveats, i.e edge cases where it might not behave as expected
2. how that actually works

**Comment by floam** [comment, marked as answer]
```console
$ printf %s\n 1 2 3 | read --list --null foo
$ set --show foo
$foo: set in global scope, unexported, with 3 elements
$foo[1]: |1|
$foo[2]: |2|
$foo[3]: |3|
```

**Reply by faho** [reply]
To expand on that:

> how that actually works

`read` normally reads until the end of the line - a `\n`. That means in `printf '%s\n' 1 2 3 | read` it will only get the first line (`1`) and then stop.

If `--list` is given, it then splits that line it has read into multiple elements on whitespace - newlines, tabs and spaces. Typically you'd do something like

```fish
echo 1 2 3 | read --list foo
```

which reads `1 2 3` and splits it into `1`, `2` and `3`.

However, because you already have lines here, you can instead read until a null-byte (`\0`) with `--null`, so the read will read all input here:

```
1
2
3
```

or, in escaped form: `1\n2\n3\n`
(because there is no \0 here anywhere)

Then it will split it on whitespace - in this case the `\n` newline characters, and split it into 1, 2 and 3.

>if there are any caveats, i.e edge cases where it might not behave as expected

Well, if there's a null-byte in your input it stops there instead of at the end of the input. Thankfully that doesn't happen a lot, because null-bytes are the unix string terminator. They can't be given in commandline arguments (e.g. if you tried passing `1\02`, the command could only read until the `1` because it has no way to know that there's more to read, and reading more than that could crash or worse if there isn't anything there), and they don't naturally occur inside of text. So, for most uses, it's safe.

The thing that was supposed to solve this was `--line`, but unfortunately it wasn't super well thought out and merged without too much review, so it can only read a set number of variables:

```
printf %s\n 1 2 3 | read --line one two three
```

But really: What you want is a command substitution. Just use a command substitution. Reading all output and putting it into a variable is what they made for. They're simple, clear, well-known. Any performance improvement you're seeing here is most likely only important in micro-benchmarks and you're talking yourself into doing awkward things in the pursuit of optimal performance that doesn't actually matter. I refer to my comments elsewhere. Just use a command substitution.

**Reply by IlanCosman** [reply]
Thank you for the explanation Fabian, it was very informative.

> Just use a command substitution.

Your wish is my command 😄
